### ISSUE TITLE
Request.WithContext() deadlock when called from handler

### ISSUE THREAD

**Issue by lende** [pkg/sftp#280]
Calling `Request.WithContext()` in our handlers (middleware) causes a deadlock. This only happens after the race condition fix in v1.8.3 (which likely solves another issue we've experienced).

What happens:

1. `fileget()` locks the given `Request` and calls the provided `FileReader` with said request
2. The `FileReader`, which is our handler (middleware) in this case, calls `Request.WithContext()`
3. `Request.WithContext()` calls `Request.copy()`, causing a deadlock

I assume this is unintended given the obvious influence from the `http`-package, where this pattern (enriching the context via middleware) is very common.

**Comment by eikenb** [comment]
Well that is just completely broken. Quick way to reproduce is to just add  `_ = r.WithContext(r.Context())` to request_example.Fileread() and then `go test -run RequestRead` and you'll get the deadlock.

Only idea off the top of my head is to split the 'state' sub-struct out of the Request and have both stored in parallel for each request (Request and RequestState or something). Basically eliminate the need for a lock on the Request passed to the handler.

**Comment by eikenb** [comment]
Another idea might work out builds on some other ongoing work I've been playing at. Related to https://github.com/pkg/sftp/pull/271#issuecomment-421157790, I've been working on moving the reader/writer/lister initialization into the Request 'factory' triggered by the appropriate OPEN packets. Once this is done, the only mutation of the Request object post-creation is updating the `lsoffset` value for Readdir requests. If we could isolate it somehow, that should eliminate the need for the global Request lock. At least it seems like it should.

**Comment by eikenb** [comment]
Sorry for the long dead period... but I'm back working on this and have had some success. Currently I have it initializing the reader/writer/lister on the open packets for both files and directories. I've also eliminated the need for changing the request.Method after the fact.

I'm about to start experimenting on what it will take to eliminate that global request lock.

**Comment by eikenb** [comment]
As far as this bug is concerned the lock issue has been fixed in my code. There are still a few uses of the request lock but they don't conflict. While I'd like to clean this up more, what I have is sufficient for this issue so I'm going to clean it up a bit and push it in a PR for preview.
